// @material/mwc-radio/mwc-radio.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/**
@license
Copyright 2018 Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { FormElement, query, customElement, property, html, observer } from '@material/mwc-base/form-element.js';
import { style } from './mwc-radio-css.js';
import { SelectionController } from './selection-controller.js';
import { ripple } from '@material/mwc-ripple/ripple-directive.js';
import MDCRadioFoundation from '@material/radio/foundation.js';
let Radio = class Radio extends FormElement {
    constructor() {
        super();
        this.checked = false;
        this.disabled = false;
        this.value = '';
        this.name = '';
        this.mdcFoundationClass = MDCRadioFoundation;
        this._selectionController = null;
        this._changeHandler = () => {
            this.checked = this.formElement.checked;
            if (this._selectionController) {
                this._selectionController.update(this);
            }
        };
        this._focusHandler = () => {
            if (this._selectionController) {
                this._selectionController.focus(this);
            }
        };
        this._clickHandler = () => {
            // Firefox has weird behavior with radios if they are not focused
            this.formElement.focus();
        };
        // Selection Controller is only needed for native ShadowDOM
        if (!window['ShadyDOM'] || !window['ShadyDOM']['inUse']) {
            this._selectionController = SelectionController.getController(this);
        }
    }
    connectedCallback() {
        super.connectedCallback();
        if (this._selectionController) {
            this._selectionController.register(this);
        }
    }
    disconnectedCallback() {
        if (this._selectionController) {
            this._selectionController.unregister(this);
        }
    }
    focusNative() {
        this.formElement.focus();
    }
    renderStyle() {
        return style;
    }
    get ripple() {
        return this.mdcRoot.ripple;
    }
    createAdapter() {
        return Object.assign({}, super.createAdapter(), { getNativeControl: () => {
                return this.formElement;
            } });
    }
    render() {
        return html `
      ${this.renderStyle()}
      <div class="mdc-radio" .ripple="${ripple()}">
        <input class="mdc-radio__native-control" type="radio" name="${this.name}" .checked="${this.checked}" .value="${this.value}"
        @change="${this._changeHandler}"
        @focus="${this._focusHandler}"
        @click="${this._clickHandler}">
        <div class="mdc-radio__background">
          <div class="mdc-radio__outer-circle"></div>
          <div class="mdc-radio__inner-circle"></div>
        </div>
      </div>`;
    }
    firstUpdated() {
        super.firstUpdated();
        if (this._selectionController) {
            this._selectionController.update(this);
        }
    }
};
__decorate([
    query('.mdc-radio')
], Radio.prototype, "mdcRoot", void 0);
__decorate([
    query('input')
], Radio.prototype, "formElement", void 0);
__decorate([
    property({ type: Boolean }),
    observer(function (checked) {
        this.mdcFoundation.setChecked(checked);
    })
], Radio.prototype, "checked", void 0);
__decorate([
    property({ type: Boolean }),
    observer(function (disabled) {
        this.mdcFoundation.setDisabled(disabled);
    })
], Radio.prototype, "disabled", void 0);
__decorate([
    property({ type: String }),
    observer(function (value) {
        this.mdcFoundation.setValue(value);
    })
], Radio.prototype, "value", void 0);
__decorate([
    property({ type: String })
], Radio.prototype, "name", void 0);
Radio = __decorate([
    customElement('mwc-radio')
], Radio);
export { Radio };
