// @material/mwc-radio/selection-controller.js
/**
@license
Copyright 2018 Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { Radio } from './mwc-radio.js';
/**
 * Unique symbol for marking roots
 */
const selectionController = Symbol('selection controller');
class SelectionSet {
    constructor() {
        this.selected = null;
        this.ordered = null;
        this.set = new Set();
    }
}
export class SelectionController {
    constructor(element) {
        this.sets = {};
        this.focusedSet = null;
        this.mouseIsDown = false;
        this.updating = false;
        element.addEventListener('keydown', (e) => this.keyDownHandler(e));
        element.addEventListener('mousedown', () => this.mousedownHandler());
        element.addEventListener('mouseup', () => this.mouseupHandler());
    }
    static getController(element) {
        const root = element.getRootNode();
        if (!root[selectionController]) {
            root[selectionController] = new SelectionController(root);
        }
        return root[selectionController];
    }
    keyDownHandler(e) {
        if (!(e.target instanceof Radio)) {
            return;
        }
        const element = e.target;
        if (!this.has(element)) {
            return;
        }
        if (e.key == 'ArrowRight' || e.key == 'ArrowDown') {
            this.next(element);
        }
        else if (e.key == 'ArrowLeft' || e.key == 'ArrowUp') {
            this.previous(element);
        }
    }
    mousedownHandler() {
        this.mouseIsDown = true;
    }
    mouseupHandler() {
        this.mouseIsDown = false;
    }
    has(element) {
        const set = this.getSet(element.name);
        return set.set.has(element);
    }
    previous(element) {
        const order = this.getOrdered(element);
        const i = order.indexOf(element);
        this.select(order[i - 1] || order[order.length - 1]);
    }
    next(element) {
        const order = this.getOrdered(element);
        const i = order.indexOf(element);
        this.select(order[i + 1] || order[0]);
    }
    select(element) {
        element.click();
    }
    /**
     * Helps to track the focused selection group and if it changes, focuses
     * the selected item in the group. This matches native radio button behavior.
     */
    focus(element) {
        // Only manage focus state when using keyboard
        if (this.mouseIsDown) {
            return;
        }
        const set = this.getSet(element.name);
        const currentFocusedSet = this.focusedSet;
        this.focusedSet = set;
        if (currentFocusedSet != set && set.selected && set.selected != element) {
            set.selected.focusNative();
        }
    }
    getOrdered(element) {
        const set = this.getSet(element.name);
        if (!set.ordered) {
            set.ordered = Array.from(set.set);
            set.ordered.sort((a, b) => a.compareDocumentPosition(b) == Node.DOCUMENT_POSITION_PRECEDING ? 1 : 0);
        }
        return set.ordered;
    }
    getSet(name) {
        if (!this.sets[name]) {
            this.sets[name] = new SelectionSet();
        }
        return this.sets[name];
    }
    register(element) {
        const set = this.getSet(element.name);
        set.set.add(element);
        set.ordered = null;
    }
    unregister(element) {
        const set = this.getSet(element.name);
        set.set.delete(element);
        set.ordered = null;
        if (set.selected == element) {
            set.selected = null;
        }
    }
    update(element) {
        if (this.updating) {
            return;
        }
        this.updating = true;
        if (element.checked) {
            const set = this.getSet(element.name);
            for (const e of set.set) {
                e.checked = (e == element);
            }
            set.selected = element;
        }
        this.updating = false;
    }
}
